### ISSUE TITLE
TcpReceiveSendGetsCanceledByDispose tests fail on Fedora 38

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#91543]
The following tests fail on Fedora 38:
```
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
```

The exceptions look like:
```
System.AggregateException: System.AggregateException : One or more errors occurred. (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null))\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)
        at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 102
at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend, Boolean ipv6Server, Boolean dualModeClient, Boolean owning) in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1032
--- End of stack trace from previous location ---
----- Inner Stack Trace #1 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #2 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #3 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #4 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #5 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #6 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #7 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #8 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
```

These tests don't fail on Fedora 37, so the behavior is possibly triggered by a change in kernel behavior.

I will investigate further when I have some time.

cc @omajid

**Comment by karelz** [comment]
@tmds please do not remove `untriaged` label. We do it only when we set milestone. Otherwise issues might slip through :(

**Comment by tmds** [comment]
> @tmds please do not remove untriaged label. We do it only when we set milestone. Otherwise issues might slip through :(

Ok, I haven't had this feedback before. I'll stop removing the label.

**Comment by directhex** [comment]
@tmds this is failing for you on x64 with Core, right?

These exact failures are plaguing our partners at IBM, for their configurations (Mono on PPC64 little endian, Mono on s390x). At the very least that seems to imply it's not a runtime-specific issue.

e.g. https://dev.azure.com/dnceng-public/public/_build/results?buildId=421005&view=results

**Comment by tmds** [comment]
> These exact failures are plaguing our partners at IBM, for their configurations (Mono on PPC64 little endian, Mono on s390x). At the very least that seems to imply it's not a runtime-specific issue.

Right. As mentioned in the initial comment, I think it's in the kernel.

Note that these tests don't fail on our CI when they run on ppc64le with RHEL 8.
And, RHEL8 has a different kernel than what CI is using here.

**Comment by tmds** [comment]
It could be different kernel bugs ...

Has this test ever passed on the public CI setup? If so, do you know when it started to fail?

**Comment by directhex** [comment]
We only have the resources to run the main branch community jobs twice a day, it looks like the problem appeared between 9c4b135ae2b1ffb5adfae8b76486cddc92995ec5 and 7f191adb868d8665ceada6f5ac5ad9150884d490 - about a 20 commit range

**Comment by directhex** [comment]
e.g. https://dev.azure.com/dnceng-public/public/_build/results?buildId=397361&view=results

**Comment by tmds** [comment]
> We only have the resources to run the main branch community jobs twice a day, it looks like the problem appeared between https://github.com/dotnet/runtime/commit/9c4b135ae2b1ffb5adfae8b76486cddc92995ec5..https://github.com/dotnet/runtime/commit/7f191adb868d8665ceada6f5ac5ad9150884d490 - about a 20 commit range

Nothing in the commit range stands out.
Did the VM stay exactly the same over this period?

I will investigate further when I upgrade to Fedora 38, which will next month or so.

**Comment by directhex** [comment]
Interesting question. According to the dpkg logs, OS updates happened on August 23rd, and September 13th - there's nothing on the OS/VM side which changed in the problem range.

**Comment by alhad-deshpande** [comment]
I tried on Ubuntu 20.04 with kernel version 5.4.0-162-generic and it worked there.

Below is the console log:
root@dotnet1:~/dotnet-ppc64le/runtime/artifacts/bin/System.Net.Sockets.Tests/Release/net7.0-unix# ../../../testhost/net7.0-Linux-Release-ppc64le/dotnet ../../../testhost/net7.0-Linux-Release-ppc64le/dotnet exec --runtimeconfig [System.Net](http://system.net/).Sockets.Tests.runtimeconfig.json --depsfile [System.Net](http://system.net/).Sockets.Tests.deps.json xunit.console.dll [System.Net](http://system.net/).Sockets.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing^C
root@dotnet1:~/dotnet-ppc64le/runtime/artifacts/bin/System.Net.Sockets.Tests/Release/net7.0-unix# ../../../testhost/net7.0-Linux-Release-ppc64le/dotnet exec --runtimeconfig [System.Net](http://system.net/).Sockets.Tests.runtimeconfig.json --depsfile [System.Net](http://system.net/).Sockets.Tests.deps.json xunit.console.dll [System.Net](http://system.net/).Sockets.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
  Discovering: [System.Net](http://system.net/).Sockets.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  [System.Net](http://system.net/).Sockets.Tests (found 1258 of 1713 test cases)
  Starting:    [System.Net](http://system.net/).Sockets.Tests (parallel test collections = on, max threads = 8)
    [System.Net](http://system.net/).Sockets.Tests.DualModeConnectToHostString.DualModeConnect_LoopbackDnsToHost_Helper [SKIP]
      Condition(s) not met: "LocalhostIsBothIPv4AndIPv6"
    [System.Net](http://system.net/).Sockets.Tests.DualModeConnectToDnsEndPoint.DualModeConnect_DnsEndPointToHost_Helper [SKIP]
      Condition(s) not met: "LocalhostIsBothIPv4AndIPv6"
    [System.Net](http://system.net/).Sockets.Tests.DualModeConnectAsync.DualModeConnectAsync_DnsEndPointToHost_Helper [SKIP]
      Condition(s) not met: "LocalhostIsBothIPv4AndIPv6"
    [System.Net](http://system.net/).Sockets.Tests.UdpClientTest.Finalize_NoExceptionsThrown [SKIP]
      Condition(s) not met: "IsPreciseGcSupported"
    [System.Net](http://system.net/).Sockets.Tests.CreateSocket.Ctor_Raw_NotSupported_ExpectedError [SKIP]
      Condition(s) not met: "NotSupportsRawSockets"

**Comment by tmds** [comment]
Seems I don't need to upgrade to Fedora 38 to investigate, as the issue is now also reproducing on my Fedora 37 system.

These are my findings.

https://github.com/dotnet/runtime/blob/9476f404988e5f18ecdf99f19d9a4cb5a7bbb219/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L218

unexpectedly gets `EBUSY` when it calls `connect` here, which causes it to call `shutdown` as a fallback:

https://github.com/dotnet/runtime/blob/9476f404988e5f18ecdf99f19d9a4cb5a7bbb219/src/native/libs/System.Native/pal_networking.c#L3072-L3077

This is covered by the `mayShutdownGraceful` case of the test:

https://github.com/dotnet/runtime/blob/9476f404988e5f18ecdf99f19d9a4cb5a7bbb219/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs#L1130-L1137

This reproduces on my Fedora 37 system with the 6.4.15-100.fc37.x86_64 kernel, but it does not reproduce on our Fedora 37 CI system which has an older 6.4.15-100.fc37.x86_64 kernel.

I will investigate further with a kernel engineer what is causing the change in behavior from the kernel side.

To unblock the CI, you can add the failing legs to this:

https://github.com/dotnet/runtime/blob/9476f404988e5f18ecdf99f19d9a4cb5a7bbb219/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs#L1034-L1038

**Comment by tmds** [comment]
> unexpectedly gets EBUSY when it calls connect here

This is caused by this change: https://github.com/torvalds/linux/commit/4faeee0cf8a5d88d63cdbc3bab124fb0e6aed08c.

It explicitly prevents connect to work when operations are on going.
While we're using it to abort things which are on going...

We prefer `connect(AF_UNSPEC)` over `shutdown(SHUT_RDWR)` because the observable effects were the same as on Windows.

We'll see the test fail on newer kernels, and kernels that back-port this change.

We should set `mayShutdownGraceful` to `true` on any Linux.

**Comment by karelz** [comment]
Fixed in main (9.0) in PR #93198 and in 8.0.x in PR #93502 and in 7.0.x in PR #93505 and in 6.0.x in PR #93554.
